### ISSUE TITLE
08-k-nearest-neighbours-winequality-red.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/Antonio27M-machine-learning#6]
Hola @Antonio27M , por aquí las issues del ejercicio de clustering

## LO QUE HAS HECHO BIEN

- **Eliminación de duplicados**: `drop_duplicates()` - buena práctica
- **MinMaxScaler**: Escalado correcto para K-Means
- **K-Means con k=6**: Valor correcto según el enunciado
- **Stratify en train_test_split**: Buena práctica para mantener proporción de clusters
- **Matriz de confusión visualizada**: Heatmap con seaborn
- **Métricas completas**: accuracy, recall, f1, precision
- **Modelo guardado**: Usas pickle correctamente
- **average='weighted' en métricas**: Correcto para multiclase

---

### 1. ERROR: El drop de columnas NO se guarda

```python
df.drop(columns=['HouseAge','AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'MedHouseVal'])
```

Esto **NO modifica** el DataFrame original porque falta `inplace=True` o reasignar:

```python
# Opción 1:
df.drop(columns=['HouseAge','AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'MedHouseVal'], inplace=True)

# Opción 2:
df = df.drop(columns=['HouseAge','AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'MedHouseVal'])
```

Aunque luego seleccionas solo las columnas que necesitas, es confuso dejar código que no hace nada.

---

### 2. ERROR: Escalas TODO el dataset ANTES del split (DATA LEAKAGE)

```python
min_max_scaler = MinMaxScaler() 
num_variables = ['MedInc', 'Latitude', 'Longitude']
min_max_features = min_max_scaler.fit_transform(df[num_variables])  # ← Escalas TODO
```

Y luego haces el split sobre los datos ya escalados. Esto es **data leakage** porque el scaler "ve" los datos de test al calcular min/max.

**La forma correcta es:**
1. Split PRIMERO
2. fit_transform en train
3. transform en test

```python
# 1. Seleccionar columnas
X = df[['MedInc', 'Latitude', 'Longitude']]

# 2. Split PRIMERO
X_train, X_test = train_test_split(X, test_size=0.2, random_state=42)

# 3. Escalar DESPUÉS
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# 4. K-Means en train
kmeans = KMeans(n_clusters=6, random_state=42, n_init=10)
kmeans.fit(X_train_scaled)
```

---

### 3. ERROR: model.score sin paréntesis

```python
model.score  # ← Esto no ejecuta nada, solo muestra el método
```

Debería ser:
```python
model.score(X_test, y_test)  # ← Con paréntesis y argumentos
```

---

### 4. Falta visualización 

- EDA sin ningún gráfico mínimo.
- No tienes ningún scatter plot para ver cómo quedaron los clusters:

```python
fig, axes = plt.subplots(1, 3, figsize=(15, 5))

# Lon vs Lat
axes[0].scatter(df_min_max['Longitude'], df_min_max['Latitude'], 
                c=df_min_max['clusters'], cmap='tab10', alpha=0.6)
axes[0].set_title('Longitude vs Latitude')
axes[0].set_xlabel('Longitude')
axes[0].set_ylabel('Latitude')

# Lat vs MedInc
axes[1].scatter(df_min_max['Latitude'], df_min_max['MedInc'], 
                c=df_min_max['clusters'], cmap='tab10', alpha=0.6)
axes[1].set_title('Latitude vs MedInc')

# Lon vs MedInc
axes[2].scatter(df_min_max['Longitude'], df_min_max['MedInc'], 
                c=df_min_max['clusters'], cmap='tab10', alpha=0.6)
axes[2].set_title('Longitude vs MedInc')

plt.tight_layout()
plt.show()
```

---

### 5. Opcional pero importante, Silhouette Score

Importas `silhouette_score` pero nunca lo usas, esta métrica en este ámbito es muy recomendable usarla:

```python
from sklearn.metrics import silhouette_score

silhouette = silhouette_score(X_train_scaled, kmeans.labels_)
print(f"Silhouette Score: {silhouette:.4f}")
```

---

### 6. Leve: Falta guardar el K-Means y el Scaler

Solo guardas el DecisionTree:

```python
# Guardar K-Means
with open('../models/kmeans_model.pkl', 'wb') as f:
    pickle.dump(kmeans, f)

# Guardar Scaler (IMPORTANTE)
with open('../models/scaler_kmeans.pkl', 'wb') as f:
    pickle.dump(min_max_scaler, f)

# Guardar clasificador
with open('../models/classifier_kmeans.pkl', 'wb') as f:
    pickle.dump(model, f)
```

Sin el scaler no podrías procesar nuevos datos correctamente.

---

### 7. Imports redundantes

Tienes imports duplicados y muchos que no usas, déjalo limpio y ordenado :) :

```python
from sklearn.cluster import KMeans  # ← Duplicado
import matplotlib.pyplot as plt     # ← Duplicado
from sklearn.datasets import make_blobs  # ← No lo usas
from yellowbrick.cluster import KElbowVisualizer  # ← No lo usas
from sklearn.metrics import davies_bouldin_score  # ← No lo usas
from sklearn.metrics import mean_squared_error, mean_absolute_percentage_error, root_mean_squared_error, r2_score  # ← Métricas de regresión, no las usas
```
---

### 8. La única conclusión que veo es una frase final. 

No hablamos de lo que tenemos ni lo que nos está dando el modelo ni analizando lo que puede ser correcto o no. La falta de graficos de ninguna clase  Me falta ver la compresión de lo que se hace y la explicación de ello para una persona que tenga un negocio y lo vea.

Ejemplos de compañeros:

<img width="1188" height="663" alt="Image" src="https://github.com/user-attachments/assets/7007c441-a770-4271-8dd0-34f60044b7a7" />

<img width="1187" height="416" alt="Image" src="https://github.com/user-attachments/assets/fa9251b7-fa05-4cb2-915f-d5b2a335d2fd" />

<img width="1192" height="653" alt="Image" src="https://github.com/user-attachments/assets/3b425c1c-e837-48c3-94ce-618a2bc45808" />

<img width="1183" height="636" alt="Image" src="https://github.com/user-attachments/assets/f60c66a0-578f-4f1d-a1bc-fdb1608643cb" />

**Antonio**, el error más grave es la falta de profundidad, y algunos errores como el **data leakage** al escalar todo el dataset antes del split cosa que hemos visto varias veces :( . Esto significa que el scaler "conoce" información del test cuando calcula los mínimos y máximos. Además, te falta toda la parte de visualización de clusters (scatter plots).
